src/x86/sse.asm

```diff
@@ -0,0 +1,510 @@
+    pmaddubsw           m2, m0
+    pmaddubsw           m3, m0
+    pmaddwd             m2, m2
+    pmaddwd             m3, m3
+    paddd               m2, m3
+    paddd               m1, m2
+
+    %define LOAD_SCALES LOAD_SCALES_8X4
+    SSE_SCALE 1, 2
+%endmacro
+
+%macro WEIGHTED_SSE_16X4_KERNEL 0
+    pmovzxbw            m0, [srcq]
+    pmovzxbw            m1, [dstq]
+    psubw               m0, m1
+    pmaddwd             m0, m0
+    pmovzxbw            m1, [srcq+src_strideq]
+    pmovzxbw            m2, [dstq+dst_strideq]
+    psubw               m1, m2
+    pmaddwd             m1, m1
+    paddd               m0, m1
+    pmovzxbw            m1, [srcq+src_strideq*2]
+    pmovzxbw            m2, [dstq+dst_strideq*2]
+    psubw               m1, m2
+    pmaddwd             m1, m1
+    pmovzxbw            m2, [srcq+src_stride3q]
+    pmovzxbw            m3, [dstq+dst_stride3q]
+    psubw               m2, m3
+    pmaddwd             m2, m2
+    paddd               m1, m2
+    paddd               m1, m0
+
+    %define LOAD_SCALES LOAD_SCALES_16X4
+    SSE_SCALE 1, 2
+%endmacro
+
+%macro WEIGHTED_SSE_32X4_KERNEL 0
+    ; Unpacking high and low results in sums that are 8 samples apart. To
+    ; correctly apply weights, two separate registers are needed to accumulate.
+    mova                m2, [srcq]
+    mova                m3, [dstq]
+    punpcklbw           m1, m2, m3
+    punpckhbw           m2, m3
+    mova                m4, [srcq+src_strideq]
+    mova                m5, [dstq+dst_strideq]
+    punpcklbw           m3, m4, m5
+    punpckhbw           m4, m5
+    pmaddubsw           m1, m0
+    pmaddubsw           m2, m0
+    pmaddubsw           m3, m0
+    pmaddubsw           m4, m0
+    pmaddwd             m1, m1
+    pmaddwd             m2, m2
+    pmaddwd             m3, m3
+    pmaddwd             m4, m4
+    ; Accumulate
+    paddd               m1, m3
+    paddd               m2, m4
+    mova                m4, [srcq+src_strideq*2]
+    mova                m5, [dstq+dst_strideq*2]
+    punpcklbw           m3, m4, m5
+    punpckhbw           m4, m5
+    mova                m6, [srcq+src_stride3q]
+    mova                m7, [dstq+dst_stride3q]
+    punpcklbw           m5, m6, m7
+    punpckhbw           m6, m7
+    pmaddubsw           m3, m0
+    pmaddubsw           m4, m0
+    pmaddubsw           m5, m0
+    pmaddubsw           m6, m0
+    pmaddwd             m3, m3
+    pmaddwd             m4, m4
+    pmaddwd             m5, m5
+    pmaddwd             m6, m6
+    paddd               m3, m5
+    paddd               m4, m6
+    paddd               m1, m3
+    paddd               m2, m4
+
+    SSE_SCALE_32X4 1, 2, 3, 4
+%endmacro
+
+%macro WEIGHTED_SSE 2 ; w, h
+%if %1 == 8
+%if %2 == 4
+; Use scale_stride's register to store src_stride3
+cglobal weighted_sse_%1x%2, 6, 7, 5, \
+        src, src_stride, dst, dst_stride, scale, \
+        src_stride3, dst_stride3
+%else
+cglobal weighted_sse_%1x%2, 6, 9, 6, \
```

Comment:
You can have C-like expressions when declaring the number of registers.
e.g. cglobal weighted_sse_%1x%2, 6, 9 + 2*(%1 == 32 ? 9 >> 1 : 2), 6 + 4*(%2 == 32), and so on.
So you can clean up the ifdeffery somewhat.

Reply:
No, this should work on yasm and ancient nasm versions as well. Its used pretty extensively in ffmpeg.
Regardless, we don't support yasm or any old versions of nasm.